File: Platforms/Windows/Reign.Video.D3D9/SamplerStateDesc.cpp
#include "pch.h"
#include "SamplerState.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	D3DTEXTUREFILTERTYPE SamplerStateDesc::Filter::get() {return filter;}

	D3DTEXTUREADDRESS SamplerStateDesc::AddressU::get() {return addressU;}
	D3DTEXTUREADDRESS SamplerStateDesc::AddressV::get() {return addressV;}
	D3DTEXTUREADDRESS SamplerStateDesc::AddressW::get() {return addressW;}

	D3DCOLOR SamplerStateDesc::BorderColor::get() {return borderColor;}
	#pragma endregion

	#pragma region Constructors
	SamplerStateDesc::SamplerStateDesc(SamplerStateTypes type)
	{
		switch (type)
		{
			case (SamplerStateTypes::Point_Wrap):
				filter = D3DTEXF_POINT;
				addressU = D3DTADDRESS_WRAP;
				addressV = D3DTADDRESS_WRAP;
				addressW = D3DTADDRESS_WRAP;
				borderColor = D3DCOLOR_ARGB(0, 0, 0, 0);
				break;

			case (SamplerStateTypes::Point_Clamp):
				filter = D3DTEXF_POINT;
				addressU = D3DTADDRESS_CLAMP;
				addressV = D3DTADDRESS_CLAMP;
				addressW = D3DTADDRESS_CLAMP;
				borderColor = D3DCOLOR_ARGB(0, 0, 0, 0);
				break;

			case (SamplerStateTypes::Linear_Wrap):
				filter = D3DTEXF_LINEAR;
				addressU = D3DTADDRESS_WRAP;
				addressV = D3DTADDRESS_WRAP;
				addressW = D3DTADDRESS_WRAP;
				borderColor = D3DCOLOR_ARGB(0, 0, 0, 0);
				break;

			case (SamplerStateTypes::Linear_Clamp):
				filter = D3DTEXF_LINEAR;
				addressU = D3DTADDRESS_CLAMP;
				addressV = D3DTADDRESS_CLAMP;
				addressW = D3DTADDRESS_CLAMP;
				borderColor = D3DCOLOR_ARGB(0, 0, 0, 0);
				break;

			default:
				Debug::ThrowError(L"SamplerStateDesc", L"Unsuported SamplerStateType");
				break;
		}
	}
	#pragma endregion
}
}
}

File: Platforms/Win32/Reign.Video.D3D9/States/SamplerStateDesc.cs
using System;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class SamplerStateDesc : SamplerStateDescI
	{
		#region Properties
		internal SamplerStateDescCom com;
		#endregion

		#region Constructors
		public static SamplerStateDesc New(SamplerStateTypes type)
		{
			return new SamplerStateDesc(type);
		}

		public SamplerStateDesc(SamplerStateTypes type)
		{
			REIGN_D3DTEXTUREFILTERTYPE filter = REIGN_D3DTEXTUREFILTERTYPE.ANISOTROPIC;
			REIGN_D3DTEXTUREADDRESS addressU = REIGN_D3DTEXTUREADDRESS.BORDER, addressV = REIGN_D3DTEXTUREADDRESS.BORDER, addressW = REIGN_D3DTEXTUREADDRESS.BORDER;
			byte r = 0, g = 0, b = 0, a = 0;

			switch (type)
			{
				case SamplerStateTypes.Point_Wrap:
					filter = REIGN_D3DTEXTUREFILTERTYPE.POINT;
					addressU = REIGN_D3DTEXTUREADDRESS.WRAP;
					addressV = REIGN_D3DTEXTUREADDRESS.WRAP;
					addressW = REIGN_D3DTEXTUREADDRESS.WRAP;
					break;

				case SamplerStateTypes.Point_Clamp:
					filter = REIGN_D3DTEXTUREFILTERTYPE.POINT;
					addressU = REIGN_D3DTEXTUREADDRESS.CLAMP;
					addressV = REIGN_D3DTEXTUREADDRESS.CLAMP;
					addressW = REIGN_D3DTEXTUREADDRESS.CLAMP;
					break;

				case SamplerStateTypes.Linear_Wrap:
					filter = REIGN_D3DTEXTUREFILTERTYPE.LINEAR;
					addressU = REIGN_D3DTEXTUREADDRESS.WRAP;
					addressV = REIGN_D3DTEXTUREADDRESS.WRAP;
					addressW = REIGN_D3DTEXTUREADDRESS.WRAP;
					break;

				case SamplerStateTypes.Linear_Clamp:
					filter = REIGN_D3DTEXTUREFILTERTYPE.LINEAR;
					addressU = REIGN_D3DTEXTUREADDRESS.CLAMP;
					addressV = REIGN_D3DTEXTUREADDRESS.CLAMP;
					addressW = REIGN_D3DTEXTUREADDRESS.CLAMP;
					break;

				default:
					Debug.ThrowError("SamplerStateDesc", "Unsuported SamplerStateType");
					break;
			}

			com = new SamplerStateDescCom(filter, addressU, addressV, addressW, r, g, b, a);
		}
		#endregion
	}
}